### PR TITLE
WIP: Add initial RIOT port

### PR DESCRIPTION
**Pull request by kaspar030**
This PR adds initial support for the [RIOT](https://github.com/RIOT-OS/RIOT) operating system.

RIOT is a nice, portable, small but feature rich OS for small embedded devices, with small meaning from 4k RAM and up. The focus is on IoT applications, with a full-featured IPv6/6LoWPAN/RPL/... network stack nearing completion. It runs on 8bit, 16bit and 32bit platforms, sporting an O(1) preemptive scheduler, fast IPC and syncronization and nicely abstracted hardware support.

This port has been PRed for integration in RIOT as https://github.com/RIOT-OS/RIOT/pull/2968. (Update: merged)

~~It cannot do much as no library has been ported.~~

TODO:
- ~~there's a bug that eats the parentheses when entering "help()" on upy's command line. But e.g., "print()" works fine. don't know why. (works fine on RIOT's native linux port)~~ (seems to be pyterm related)
- ~~the garbage collector doesn't consider register contents, so it will probably not work correctly~~ fixed
- there needs to be either a mapping for riot C libs -> micropython libs or an adaption of micropythons "pyb." lib

This PR contains some cleanup commits:
- one for accepting both 0xa and 0xd as newline for readline, needed to support our native port's console input
- one that fixes obobject.c when not using `MICROPY_CPYTHON_COMPAT` (a .locals_dict had a broken definition when setting that flag)
- one that changes a function pointer from void\* to void(*)(void) to make it ISO C
- one to make mp_binary_get_size unsigned as otherwise later comparisons trigger a sign-compare warning

### CODE REVIEW

**Comment by kaspar030** [comment]
IMHO you guys should consider using RIOT as your HAL as to not have to reimplement hardware support for every new platform, but that's probably a veeeeery biased opinion. ;)

File: lib/mp-readline/readline.c [hunk 1]
```diff
@@ -35,8 +35,8 @@
 #include MICROPY_HAL_H
 #endif
 
-#if 0 // print debugging info
-#define DEBUG_PRINT (1)
+#if 1 // print debugging info
+#define DEBUG_PRINT (0)
```
Comment (OlegHahm) [inline lib/mp-readline/readline.c]: I guess this shouldn't go into this PR.

Comment (kaspar030) [inline lib/mp-readline/readline.c]: ups, changed back the wrong way.

**Comment by pfalcon** [comment]
> IMHO you guys should consider using RIOT as your HAL as to not have to reimplement hardware support for every new platform, but that's probably a veeeeery biased opinion. ;)

I guess every RTOS could say the same, and that's a good reason to keep MicroPython core RTOS-free, while inviting as many as possible of them to participate on equal conditions. But RIOT is definitely something interesting which I for example had in my list for more detailed look for a long time. Guess this will be a nice chance to actually look a bit deeper into it ;-).

**Comment by kaspar030** [comment]
> I guess every RTOS could say the same, and that's a good reason to keep MicroPython core RTOS-free, while inviting as many as possible of them to participate on equal conditions.

You're probably right. ;)

Unfortunately RIOT is currently missing a file system, but I guess micropython will trigger development in that direction...

**Comment by danicampora** [comment]
> This PR adds initial support for the RIOT operating system.

Nice!

> Unfortunately RIOT is currently missing a file system, but I guess micropython will trigger development in that direction...

We are already using FatFS in several ports, have you considered it?

**Comment by kaspar030** [comment]
> We are already using FatFS in several ports, have you considered it?

Actually, as I saw it in micropython, I got the same idea. So it works fine for you guys?

(We've had not-so-nice experiences with FAT in an early version of RIOT. Might have been the implementation.)

**Comment by danicampora** [comment]
> Actually, as I saw it in micropython, I got the same idea. So it works fine for you guys?

Works like a charm ;-)

> (We've had not-so-nice experiences with FAT in an early version of RIOT. Might have been the implementation.)

What issues did you have in the past?

**Comment by dpgeorge** [comment]
Certainly we can merge the small changes to the core.

But a whole new top-level directory which at the moment is almost the same as minimal/... that's hard to do.

@kaspar030 are you able to maintain this port?

**Comment by kaspar030** [comment]
> What issues did you have in the past?

It's been a while, but I remember that it was a nightmare to deal with all the variants and inconsistencies of FAT. But that was at a time when other things were more pressing, so this is no reason to not look into fatfs.

**Comment by kaspar030** [comment]
> @kaspar030 are you able to maintain this port?

I think so. Actually I'm pretty sure that my fellow RIOT friends will help fill in the pyb.* stuff and keep this updated.

I can keep this in my micropython fork for now until it has enough code to justify the top-level dir.

**Comment by kaspar030** [comment]
BTW, great work so far with mircopython! Thanks a lot!

**Comment by danicampora** [comment]
> It's been a while, but I remember that it was a nightmare to deal with all the variants and inconsistencies of FAT.

So you guys were trying to write your own FAT implementation? Why reinvent the wheel? FatFs already takes care of all the dirty details, and it's AFAIK the most popular open source FAT implementation.

**Comment by kaspar030** [comment]
@danicampora I don't remember... That was 2008, I was still a student and just WTFing over the commited code... I guess fatfs didn't fit somehow back than, or the reserach was very bad.

**Comment by danicampora** [comment]
> I don't remember... That was 2008
OK, I didn't know this project had that much history.

**Comment by kaspar030** [comment]
> OK, I didn't know this project had that much history.
Well, it had at least two different names until it came to new life as RIOT in 2013. The core was started in 2008.

I just realized that the ominous "help()" problem I'm experiencing seems to be with our terminal (pyterm), using picocom, it works fine.

**Comment by OlegHahm** [comment]
> So you guys were trying to write your own FAT implementation? Why reinvent the wheel?

No, we were using the implementation from ChaN: http://elm-chan.org/fsw/ff/00index_e.html

**Comment by dpgeorge** [comment]
We use the ChaN fatfs library.  So far we have had very few issues with it (I think the only issue we had was recently fixed in the latest version).

**Comment by OlegHahm** [comment]
I cannot say much about the problems with FAT back then, because I never had a use case for it, but maybe it just improved during the last seven years. ;)

**Comment by dpgeorge** [comment]
Do you have some info on how to build a complete binary image for a real board, such as the STM32F4 Discovery board?

**Comment by kaspar030** [comment]
You'd checkout the branch, got to test/pkg_micropython, then "BOARD=stm32f4discovery make clean all flash", then "BOARD=stm32f4discovery make term". Maybe you also have to specify the serial port the board is connected to, e.g., "BOARD=stm32f4discovery PORT=/dev/ttyUSB1 make term".

Our default terminal does a lot of what mp-readline does, so expect the cli to be flakey.

**Comment by dpgeorge** [comment]
I can confirm that it works on an STM32F4 Discovery board.  Note that there is no USB driver in RIOT, rather the serial console is via UART2 on PA2=TX, PA3=RX.

**Comment by dpgeorge** [comment]
I merged 3 of the commits making changes to the core (to obj.h, binary.[ch] and objobject.c), which should fix your warnings.

Regarding the change to readline, allowing '\n' to be equivalent to '\r': we could merge this, I think it doesn't hurt as there would be no other use for '\n' except to make a newline (is it ever the case that you can receive \n and \r to mean 1 newline?).

**Comment by pfalcon** [comment]
> Regarding the change to readline, allowing '\n' to be equivalent to '\r': we could merge this, I think it doesn't hurt as there would be no other use for '\n' except to make a newline (is it ever the case that you can receive \n and \r to mean 1 newline?).

Yes, I would like to know more details why this was needed too - I also think it won't hurt, but then sooner or later we may receive bug report of pressing Enter giving an extra empty line with someone using some weird terminal app, then we need to assess which fix is more important...

**Comment by dpgeorge** [comment]
Such a change to the newline handling code would actually be useful also to the unix port when it uses mp-readline.  When I made #1263 I was almost going to allow '\n' to be a newline, but instead did this: https://github.com/micropython/micropython/pull/1263/files#diff-15f11433aab2b0271dd5463445d67282R110

**Comment by kaspar030** [comment]
> Yes, I would like to know more details why this was needed too

Well, we have this "native port", which is basically a RIOT instance running as Linux or OSX user space process. That is tremendously useful for debugging. But as it's "serial console" is the host system's stdin, it receives a '\n' as newline character. So we're accepting both '\r' and '\n' as newline char.

We could also go for the way it's done in #1263, your call.

**Comment by kaspar030** [comment]
Please excuse the lack of updates here. We're trying to get a release out, I will find time to update this afterwards (couple of weeks).

**Comment by jgriessen** [comment]
Is there a plan to let RIOT code be importable to work with micropython, or has this been dropped?

**Comment by dpgeorge** [comment]
> Is there a plan to let RIOT code be importable to work with micropython, or has this been dropped?

We would need @kaspar030 to work on it.  Otherwise it'll be considered dropped (for now).

**Comment by kaspar030** [comment]
I'll continue working on this (and I'm hoping to get GSOC'16 support).

**Comment by jgriessen** [comment]
"I'll continue working on this (and I'm hoping to get GSOC'16 support)."  Would you tell us more about how you can see RIOT working as a RTOS along with micropython?  How would the different ways of handling hardware platforms be reconciled?  Mapping one to the other?  I'd like mapping -- then you can imagine importing just a part of RIOT that you want and not go over the RAM limits.  Maybe the mapping would make sense by an external script you run to import code.  John Griessen

**Comment by kaspar030** [comment]
> Would you tell us more about how you can see RIOT working as a RTOS along with micropython?

Well, RIOT has quite nice C API's for hardware access, networking, ..., which basically abstract the underlying MCU. My plan is to port the micropython libraries to use those API's as backend.
That way, any MCU supported by RIOT (and which is powerful enough) will get micropython support.

I don't quite get what you mean by "mapping".

**Comment by jgriessen** [comment]
> @jgriessen I don't quite get what you mean by "mapping".

I am thinking of using the riot C API's for hardware access in python style.  I was thinking of translating those into C language calls from micropython. Probably you mean the same thing when you say, "port the micropython libraries to use those API's as backend

**Comment by kaspar030** [comment]
Alright, it's been a while, but I'll finally have time to bring this to a usable state.
The latest commit just updates the "draft" port to current master of both RIOT and micropython.

On the micropython side I didn't have to change anything (just cleaned up a bit), so thanks a lot for keeping your codebase that stable!

**Comment by jgriessen** [comment]
So, How does it work?  Is there a guide?  
"plan is to port the micropython libraries to use those API's as backend"
Are there particular libraries you have ported?

**Comment by kaspar030** [comment]
> So, How does it work? Is there a guide?

Check out RIOT (including the RIOT-side PR: https://github.com/RIOT-OS/RIOT/pull/2968), run ```make -C examples/pkg_micropython all```, should get you the "native" version where RIOT runs as Unix process.

To run on a real board, something like
    
    BOARD=stmf4discovery make -C examples/pkg_micropython flash

 ... should get you started.

> > "plan is to port the micropython libraries to use those API's as backend"
> Are there particular libraries you have ported?

Nothing yet.

**Comment by jgriessen** [comment]
So, if I build for BOARD=stmf4discovery will this be a combination of micropython with a RIOT real time event/interrupt control loop?  I'm not understanding what will be there before any libraries are ported.

**Comment by kaspar030** [comment]
> So, if I build for BOARD=stmf4discovery will this be a combination of micropython with a RIOT real time event/interrupt control loop? I'm not understanding what will be there before any libraries are ported.

If built like that, there will be the RIOT main thread running the micropython interpreter, first executing boot.py (which gets included during compilation, RIOT doesn't have a file system, yet), then the micropython  REPL, without libraries.

This might seem super boring to anyone who has actually used micropython. ;) But this is work in progress, and allows testing of the basic interpreter.

Currently I'm figuring out how micropython's garbage collection works, in order to set it up for RIOT with minimal architecture dependent code.

The next step will be adding libraries. Without them, you're right, there's not much.

**Comment by jgriessen** [comment]
On 07/18/2016 10:50 AM, Kaspar Schleiser wrote:
> This might seem super boring to anyone who has actually used micropython. ;) But this is work in progress

Thanks -- it's not boring.  I will try to help you with testing.  What are the memory requirements with RIOT + Micropython?

Would it still owrk to use a STM32F401RE with 512k flash?

**Comment by kaspar030** [comment]
> Thanks -- it's not boring.  I will try to help you with testing.  

Appreciated!

> What are the memory requirements with RIOT + Micropython?
> Would it still owrk to use a STM32F401RE with 512k flash?

Definitely. This basic test application is using 12kB RAM (including 8kB heap for micropython) and ~85kB of flash, compiled for a Cortex-M0+ board. Of course those numbers will go up as we add features, but the STM32F401RE (with 96kB RAM?) should be able to run networked applications.

I must admit I have no clue how much memory micropython actually needs. What are usable values for it's heapsize, and how much stack does the interpreter need? I know the former depends, so some rough estimations would be nice. And does the needed stack space have an upper bound?

**Comment by kaspar030** [comment]
I'm getting a segmentation fault:

```
Program received signal SIGSEGV, Segmentation fault.
0x0804b5dd in find_qstr (q=14982) at /home/kaspar/src/riot.tmp/tests/pkg_micropython/bin/pkg/native/micropython/riot/../py/qstr.c:101
101                 return pool->qstrs[q - pool->total_prev_len];
(gdb) bt
#0  0x0804b5dd in find_qstr (q=14982) at /home/kaspar/src/riot.tmp/tests/pkg_micropython/bin/pkg/native/micropython/riot/../py/qstr.c:101
#1  0x0804b934 in qstr_hash (q=14982) at /home/kaspar/src/riot.tmp/tests/pkg_micropython/bin/pkg/native/micropython/riot/../py/qstr.c:190
#2  0x08060617 in mp_obj_str_get_hash (self_in=0xea1a)
    at /home/kaspar/src/riot.tmp/tests/pkg_micropython/bin/pkg/native/micropython/riot/../py/objstr.c:1988
#3  0x08058a29 in mp_obj_hash (o_in=0xea1a) at /home/kaspar/src/riot.tmp/tests/pkg_micropython/bin/pkg/native/micropython/riot/../py/obj.c:160
#4  0x08058420 in mp_map_lookup (map=0x807d9d4 <mp_state_ctx+52>, index=0xea1a, lookup_kind=MP_MAP_LOOKUP)
    at /home/kaspar/src/riot.tmp/tests/pkg_micropython/bin/pkg/native/micropython/riot/../py/map.c:203
#5  0x0805618f in mp_load_global (qst=14982) at /home/kaspar/src/riot.tmp/tests/pkg_micropython/bin/pkg/native/micropython/riot/../py/runtime.c:134
#6  0x08056165 in mp_load_name (qst=14982) at /home/kaspar/src/riot.tmp/tests/pkg_micropython/bin/pkg/native/micropython/riot/../py/runtime.c:128
#7  0x08064ec8 in mp_execute_bytecode (code_state=<optimized out>, inject_exc=<optimized out>)
    at /home/kaspar/src/riot.tmp/tests/pkg_micropython/bin/pkg/native/micropython/riot/../py/vm.c:255
#8  0x0805ad29 in fun_bc_call (self_in=0x807ba10 <mp_heap+176>, n_args=0, n_kw=0, args=0x0)
    at /home/kaspar/src/riot.tmp/tests/pkg_micropython/bin/pkg/native/micropython/riot/../py/objfun.c:237
#9  0x080569cd in mp_call_function_n_kw (fun_in=0x807ba10 <mp_heap+176>, n_args=0, n_kw=0, args=0x0)
    at /home/kaspar/src/riot.tmp/tests/pkg_micropython/bin/pkg/native/micropython/riot/../py/runtime.c:560
#10 0x08056951 in mp_call_function_0 (fun=0x807ba10 <mp_heap+176>)
    at /home/kaspar/src/riot.tmp/tests/pkg_micropython/bin/pkg/native/micropython/riot/../py/runtime.c:534
#11 0x0804966e in mp_do_str (
    src=0x807b1e0 <boot_py> "print(\"boot.py: Micro Python RIOT port initialized.\")\n\nfor i in range(0, 5):\n    print(\"Iteration\", i)\n    x = []\n    for j in range(0, 100):\n        x.append(str(j))\n\n    for n in x:\n        print(n)"..., len=201) at mp_riot.c:27
#12 0x080495c4 in main () at main.c:44
```

When executing a simple gc test script:

```
print("boot.py: Micro Python RIOT port initialized.")

for i in range(0, 5):
    print("Iteration", i)
    x = []
    for j in range(0, 100):
        x.append(str(j))

    for n in x:
        print(n)
```

The garbage collector doesn't consider registers (yet), could that lead to qstrings (or any object) getting garbage collected while still in use?

**Comment by kaspar030** [comment]
> The garbage collector doesn't consider registers (yet), could that lead to qstrings (or any object) getting garbage collected while still in use?

Solved it.

**Comment by jedie** [comment]
The link in the initial Post is wrong here:

> This port has been PRed for integration in RIOT as #2968 (in riot's github repo).

The right Pull Request @ RIOT is -> https://github.com/RIOT-OS/RIOT/pull/2968

**Comment by kaspar030** [comment]
> The right Pull Request @ RIOT is -> [RIOT-OS/RIOT#2968](https://github.com/RIOT-OS/RIOT/pull/2968)

Thanks, fixed!

We've merged this into RIOT. Working on getting at least basic peripherals in the machine module working.

**Comment by dpgeorge** [comment]
@kaspar030 it's great you have integrated MicroPython into RIOT, and I encourage you to maintain that integration.  But I think it makes much more sense for you to have that uPy-RIOT port in the RIOT code base, and built as part of your CI, etc.  Rather than having it hosted here.

We are happy to make changes to core MicroPython to facilitate your out-of-tree port, so please to raise any PRs if you need changes like that made.

I'll close this PR, but thanks for all your efforts!

**Comment by OlegHahm** [comment]
May I ask why you think that RIOT support should not be integrated into upstream micropython? I see that already many different platforms and OS are supported, so basically RIOT would be just another candidate in the list. Testing its support for various supported hardware platforms in RIOT would still happen in the RIOT CI.

**Comment by jimmo** [comment]
@OlegHahm I agree it would be nice to be able to take this on, but it's not that simple. Adding an upstream port requires:
* An extensive and code review to initially merge it
* Ongoing maintenance to keep it up to date as MicroPython evolves
* Maintainer experience and knowledge with the underlying port/hardware/os (in this case, RIOT) to review future PRs
* Possibly providing pre-built downloads / firmware images like we do for other ports

Right now we just don't have the capacity.

**Pull request closed without merging**